### PR TITLE
[GENX]: Add barrier operation

### PR DESCRIPTION
**Pull request by etiotto**
Add `genx.barrier` operation to the GENX dialect.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 1]
```diff
@@ -72,33 +72,17 @@ class GENX_Op<string mnemonic, list<Trait> traits = []> :
...
+  : GENX_Op<mnemonic, !listconcat(traits, [Pure])>, 
+    Results<(outs LLVM_Type:$res)>, Arguments<(ins)> {
+  string llvmBuilder = ""
+    # "  llvm::Type *retType = builder.getInt64Ty();\n"
+    # "  llvm::Type *argType = builder.getInt32Ty();\n"
+    # "  llvm::Value *arg = llvm::ConstantInt::get(argType, " # arg # ");\n"
+    # "  $res = createDeviceFunctionCall(builder, \"" # fnName # "\",\n"
+    # "                             retType, { argType }, { arg });";
 
-class GENX_SpecialRegisterOp<string mnemonic, list<Trait> traits = []> :
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:90]: `GENX_SpecialRegisterOp` is not used

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -25,22 +25,19 @@ using namespace mlir;
 using namespace mlir::LLVM;
 using mlir::LLVM::detail::createIntrinsicCall;
 
-// Create a call to GENX Device function
-// Currently this routine will work only for calling GENX functions that
-// take a single int32 argument.
+// Create a call to SPIR device function.
 static llvm::Value *createDeviceFunctionCall(llvm::IRBuilderBase &builder,
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:29]: Generalize to allow passing in the function signature (retType + argument types) and more than one argument.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 3]
```diff
@@ -119,24 +103,42 @@ def GENX_BlockIdZOp : GENX_DeviceFunctionOp<"workgroup.id.z",
...
                         "_Z12get_global_sizej", 1>;
-
 def GENX_GridDimZOp : GENX_DeviceFunctionOp<"grid.dim.z",
                         "_Z12get_global_sizej", 2>;
 
+//===----------------------------------------------------------------------===//
+// Synchronization
+//===----------------------------------------------------------------------===//
+
+def NVVM_Barrier0Op : GENX_Op<"barrier"> {
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Typo: NVVM_Barrier0Op -> GENX_BarrierOp

Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: done

File: mlir/lib/Conversion/GPUToGENX/GPUToGENX.td [hunk 4]
```diff
@@ -17,4 +17,6 @@ include "mlir/IR/PatternBase.td"
 include "mlir/Dialect/GPU/IR/GPUOps.td"
 include "mlir/Dialect/LLVMIR/GENXOps.td"
 
+def : Pat<(GPU_BarrierOp), (GENX_BarrierOp)>;
```
Comment (whitneywhtsang) [inline mlir/lib/Conversion/GPUToGENX/GPUToGENX.td:20]: what is this for?

**Review by pengtu — APPROVED** [review]

File: mlir/lib/Conversion/GPUToGENX/GPUToGENX.td (hunk 4 above)
Comment (etiotto) [reply mlir/lib/Conversion/GPUToGENX/GPUToGENX.td:20]: This is a conversion pattern that allows lowering a `gpu.barrier` operation into a `genx.barrier` operation.

**Pull request merged**